src/System.Threading.Tasks/tests/Task/TaskStatusTest.cs

```diff
@@ -172,11 +172,21 @@ internal void RealRun()
                         {
                             //we may have reach this point too soon, let's keep spinning until the status changes.
                             while (_task.Status == TaskStatus.Running)
-                                ;
+                            {
+                                Task.Delay(1).Wait();
+                            }
+
                             //
-                            // If we're still waiting for children our Status should reflect so
+                            // If we're still waiting for children our Status should reflect so. For this verification, read the
+                            // parent task's status before the child task's status to make the child task's status more recent,
+                            // since the child task may complete during the status reads.
                             //
-                            if (_task.Status != TaskStatus.WaitingForChildrenToComplete)
+                            TaskStatus taskStatus = _task.Status;
+                            Interlocked.MemoryBarrier();
+                            TaskStatus childTaskStatus = _childTask.Status;
+                            if (taskStatus != TaskStatus.WaitingForChildrenToComplete &&
+                                childTaskStatus != TaskStatus.RanToCompletion &&
+                                childTaskStatus != TaskStatus.Faulted)
```

Comment:
Do we need to consider cancellation of the child task (I've not looked at the test to see if that's possible). Maybe instead just:
```C#
if (_task.Status != TaskStatus.WaitForChildrenToComplete &&
    !_childTask.IsCompleted)
{
    ...
}
```
I don't think the memory barrier should be necessary, as the implementation should be reading the states with volatile reads which will prevent reordering of those reads with one another.  But if you think it's necessary, ok.